Public/Src/Sandbox/Windows/DetoursServices/DetouredFunctions.cpp

```diff
@@ -1778,6 +1778,235 @@ NTSTATUS NTAPI Detoured_ZwSetInformationFile(
+/// the current working directory and environment.
+BOOL WINAPI InjectShim(
+    _In_opt_    LPCWSTR               lpApplicationName,
+    _Inout_opt_ LPWSTR                lpCommandLine,
+    _In_opt_    LPSECURITY_ATTRIBUTES lpProcessAttributes,
+    _In_opt_    LPSECURITY_ATTRIBUTES lpThreadAttributes,
+    _In_        BOOL                  bInheritHandles,
+    _In_        DWORD                 dwCreationFlags,
+    _In_opt_    LPVOID                lpEnvironment,
+    _In_opt_    LPCWSTR               lpCurrentDirectory,
+    _In_        LPSTARTUPINFOW        lpStartupInfo,
+    _Out_       LPPROCESS_INFORMATION lpProcessInformation)
+{
+    // Create a final buffer for the original command line.
+    size_t fullCmdLineSizeInChars =
+        (lpApplicationName != nullptr ? wcslen(lpApplicationName) : 0) +
+        (lpCommandLine != nullptr ? wcslen(lpCommandLine) : 0) +
+        4;  // Quotes and space and trailing null
+    wchar_t *fullCommandLine = new wchar_t[fullCmdLineSizeInChars];
+    if (fullCommandLine == nullptr)
+    {
+        Dbg(L"Failure running AnyBuild shim process - failed to allocate buffer of size %d.", fullCmdLineSizeInChars * sizeof(WCHAR));
+        SetLastError(ERROR_OUTOFMEMORY);
+        return FALSE;
+    }
+
+    fullCommandLine[0] = L'\0';
+    if (lpApplicationName != nullptr)
+    {
+        fullCommandLine[0] = L'"';
+        wcscpy_s(fullCommandLine + 1, fullCmdLineSizeInChars, lpApplicationName);
+        wcscat_s(fullCommandLine, fullCmdLineSizeInChars, L"\" ");
+    }
+    if (lpCommandLine != nullptr)
+    {
+        wcscat_s(fullCommandLine, fullCmdLineSizeInChars, lpCommandLine);
+    }
+
+    Dbg(L"Injecting AnyBuild shim '%s' for process command line '%s'", g_substituteProcessExecutionShimPath, fullCommandLine);
+    BOOL rv = Real_CreateProcessW(
+        /*lpApplicationName:*/ g_substituteProcessExecutionShimPath,
+        /*lpCommandLine:*/ fullCommandLine,
+        lpProcessAttributes,
+        lpThreadAttributes,
+        bInheritHandles,
+        dwCreationFlags,
+        lpEnvironment,
+        lpCurrentDirectory,
+        lpStartupInfo,
+        lpProcessInformation);
+
+    delete[] fullCommandLine;
+    return rv;
+}
+
+// https://stackoverflow.com/questions/216823/whats-the-best-way-to-trim-stdstring
+inline const wchar_t *trim_start(const wchar_t *str)
+{
+    while (wmemchr(L" \t\n\r", *str, 4))  ++str;
+    return str;
+}
+inline const wchar_t *trim_end(const wchar_t *end)
+{
+    while (wmemchr(L" \t\n\r", end[-1], 4)) --end;
+    return end;
+}
+inline std::wstring trim(const wchar_t *buffer, size_t len) // trim a buffer (input?)
+{
+    return std::wstring(trim_start(buffer), trim_end(buffer + len));
+}
+inline void trim_inplace(std::wstring& str)
+{
+    str.assign(trim_start(str.c_str()),
+        trim_end(str.c_str() + str.length()));
+}
+
+// Returns new value for lpCommandLine pointing to the remainder of the string

```

Comment:
>// Returns  [](start = 0, length = 11)

I feel that this shim logic needs to go to a separate class or util class.

Reply:
Added new SubstituteProcessExecution.cpp, .h